src/collections/collections.md

```diff
@@ -0,0 +1,258 @@
+## Using `heapless`
+
+`heapless` requires no setup as its collections don't depend on a global memory
+allocator. Just `use` its collections and proceed to instantiate them:
+
+``` rust
+extern crate heapless; // v0.4.x
+
+use heapless::Vec;
+use heapless::consts::*;
+
+#[entry]
+fn main() -> ! {
+    let mut xs: Vec<_, U8> = Vec::new();
+
+    xs.push(42).unwrap();
+    assert_eq!(xs.pop(), Some(42));
+}
+```
+
+You'll note two differences between these collections and the ones in `alloc`.
+
+First, you have to declare upfront the capacity of the collection. `heapless`
+collections never reallocate and have fixed capacities; this capacity is part of
+the type signature of the collection. In this case we have declared that `xs`
+has a capacity of 8 elements that is the vector can, at most, hold 8 elements.
+This is indicated by the `U8` (see [`typenum`]) in the type signature.
+
+[`typenum`]: https://crates.io/crates/typenum
+
+Second, the `push` method, and many other methods, return a `Result`. Since the
+`heapless` collections have fixed capacity all operations that insert elements
+into the collection can potentially fail. The API reflects this problem by
+returning a `Result` indicating whether the operation succeeded or not. In
+contrast, `alloc` collections will reallocate themselves on the heap to increase
+their capacity.
+
+As of version v0.4.x all `heapless` collections store all their elements inline.
+This means that an operation like `let x = heapless::Vec::new();` will allocate
+the collection on the stack, but it's also possible to allocate the collection
+on a `static` variable, or even on the heap (`Box<Vec<_, _>>`).
+
+## Trade-offs
+
+Keep these in mind when choosing between heap allocated, relocatable collections
+and fixed capacity collections.
+
+### Out Of Memory and error handling
+
+With heap allocations Out Of Memory is always a possibility and can occur in
+any place where a collection may need to grow: for example, all
+`alloc::Vec.push` invocations can potentially generate an OOM condition. Thus
+some operations can *implicitly* fail. Some `alloc` collections expose
+`try_reserve` methods that let you check for potential OOM conditions when
+growing the collection but you need be proactive about using them.
+
+If you exclusively use `heapless` collections and you don't use a memory
+allocator for anything else then an OOM condition is impossible. Instead, you'll
+have to deal with collections running out of capacity on a case by case basis.
+That is you'll have deal with *all* the `Result`s returned by methods like
+`Vec.push`.
+
+OOM failures can be harder to debug than say `unwrap`-ing on all `Result`s
+returned by `heapless::Vec.push` because the observed location of failure may
+*not* match with the location of the cause of the problem. For example, even
+`vec.reserve(1)` can trigger an OOM if the allocator is nearly exhausted because
+some other collection was leaking memory (memory leaks are possible in safe
```

Comment:
Out of curiosity + lack of knowledge: How would safe rust code, which does produce a memory leak, look like? 
Or was this note under the premiss, that safe rust code interacts with (e.g. collections) which might contain unsafe code with memory leaks?

Reply:
> How would safe rust code, which does produce a memory leak, look like?

`mem::forget(vec![1])` leaks memory and it's safe. `mem::forget` used to be `unsafe` until we noticed it was also possible to leak memory by creating reference cycles using the safe API of the `Rc` abstraction.

The day we made `forget` safe we also declared that depending on destructors running for memory safety is an anti-pattern (see the old fork-join API that was in the standard library; it never made into stable and was removed). This chain of events is known as the "leakpocalyse".